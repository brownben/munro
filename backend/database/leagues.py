import sqlite3


def leagueToJSON(league):
    # Convert SQL output to JSON Object
    if (league):
        return {
            'name': league[0],
            'website': league[1],
            'logo': league[2],
            'coordinator': league[3],
            'scoringMethod': league[4],
            'numberOfCountingEvents': league[5],
            'courses': league[6].split(','),
            'moreInformation': league[7]
        }
    else:
        return False


# Set Up Database
connection = sqlite3.connect('./databaseFiles/2.db')
cursor = connection.cursor()
cursor.execute('''
    CREATE TABLE IF NOT EXISTS leagues (
        name TEXT NOT NULL,
        website TEXT,
        logo TEXT,
        coordinator TEXT,
        scoringMethod TEXT NOT NULL,
        numberOfCountingEvents NUMBER,
        courses TEXT,
        moreInformation TEXT,
        PRIMARY KEY (name),
        UNIQUE (name))''')
connection.commit()
connection.close()

# League Database Functions


def createLeague(name, website, logo, coordinator, scoringMethod, noOfEvents, courses, moreInfo):
    courses = courses.replace(' ', '')
    connection = sqlite3.connect('./databaseFiles/2.db')
    cursor = connection.cursor()
    cursor.execute('INSERT INTO leagues (name,website,logo,coordinator,scoringMethod,numberOfCountingEvents, courses, moreInformation) VALUES (?,?,?,?,?,?,?,?)',
                   (name, website, logo, coordinator, scoringMethod, noOfEvents, courses, moreInfo))
    connection.commit()
    connection.close()


def updateLeague(oldName, name, website, logo, coordinator, scoringMethod, noOfEvents, courses, moreInfo):
    courses = courses.replace(' ', '')
    connection = sqlite3.connect('./databaseFiles/2.db')
    cursor = connection.cursor()
    cursor.execute('''
        UPDATE leagues
        SET name=?,website=?,logo=?,coordinator=?,scoringMethod=?,numberOfCountingEvents=?, courses=?,moreInformation=?
        WHERE name=?''', (name, website, logo, coordinator, scoringMethod, noOfEvents, courses, moreInfo, oldName))
    connection.commit()
    connection.close()


def deleteLeague(name):
    connection = sqlite3.connect('./databaseFiles/2.db')
    cursor = connection.cursor()
    cursor.execute('DELETE FROM leagues WHERE name=?', (name,))
    connection.commit()
    connection.close()


def findLeague(name):
    connection = sqlite3.connect('./databaseFiles/2.db')
    cursor = connection.cursor()
    result = cursor.execute('''
        SELECT name,website,logo,coordinator,scoringMethod,numberOfCountingEvents,courses, moreInformation
        FROM leagues
        WHERE name=?''', (name,)).fetchone()
    connection.close()
    json = leagueToJSON(result)
    if(json):
        json['numberOfEvents'] = getNumberOfEventsInLeague(json['name'])
    return json


def getAllLeagues():
    connection = sqlite3.connect('./databaseFiles/2.db')
    cursor = connection.cursor()
    result = cursor.execute('''
        SELECT name,website,logo,coordinator,scoringMethod,numberOfCountingEvents, courses, moreInformation
        FROM leagues
        ORDER BY name ASC''').fetchall()
    connection.close()
    leagues = list(map(leagueToJSON, result))
    for league in leagues:
        league['numberOfEvents'] = getNumberOfEventsInLeague(league['name'])
    return leagues


def deleteAllLeagues():
    connection = sqlite3.connect('./databaseFiles/2.db')
    cursor = connection.cursor()
    cursor.execute('DELETE FROM leagues')
    connection.commit()
    connection.close()


def getNumberOfEventsInLeague(name):
    connection = sqlite3.connect('./databaseFiles/2.db')
    cursor = connection.cursor()
    result = cursor.execute('''
        SELECT COUNT(events.name)
        FROM events
        WHERE events.league=?''', (name,)).fetchall()
    connection.close()
    return result[0][0]
