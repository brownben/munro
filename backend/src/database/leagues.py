from typing import Any, Iterable, Optional

from piccolo.query.methods.select import Count

from ..schemas import League, LeagueClass
from .tables import CompetitorPool as CompetitorPoolTable
from .tables import League as LeagueTable
from .tables import LeagueClass as LeagueClassTable
from .tables import LeagueGroup as LeagueGroupTable

league_fields = (
    LeagueTable.name,
    LeagueTable.tagline,
    LeagueTable.year,
    LeagueTable.coordinator,
    LeagueTable.website,
    LeagueTable.more_information,
    LeagueTable.visible,
    LeagueTable.scoring_method,
    LeagueTable.number_of_counting_events,
    LeagueTable.competitor_pool,
)

league_class_fields = (
    LeagueClassTable.name,
    LeagueClassTable.league,
    LeagueClassTable.standard_course,
    LeagueClassTable.age_class_filter,
    LeagueClassTable.club_filter,
    LeagueClassTable.number_of_counting_events,
)

league_group_fields = (
    LeagueGroupTable.name,
    LeagueGroupTable.league,
    LeagueGroupTable.min,
    LeagueGroupTable.max,
)


def as_league(record: dict[str, Any]) -> Optional[League]:
    if not record:
        return None

    return League.parse_obj(record)


def as_league_class(record: dict[str, Any]) -> Optional[LeagueClass]:
    if not record:
        return None

    return LeagueClass.parse_obj(record)


class Leagues:
    @staticmethod
    async def create(league: League) -> bool:
        if not league.competitor_pool:
            await CompetitorPoolTable.insert(
                CompetitorPoolTable(name=league.name)
            ).run()
            league.competitor_pool = league.name

        await LeagueTable.insert(
            LeagueTable(
                name=league.name,
                tagline=league.tagline,
                year=league.year,
                coordinator=league.coordinator,
                website=league.website,
                more_information=league.more_information,
                visible=league.visible,
                scoring_method=league.scoring_method,
                number_of_counting_events=league.number_of_counting_events,
                competitor_pool=league.competitor_pool,
            )
        ).run()

        return True

    @staticmethod
    async def update(name: str, league: League) -> bool:
        if league.name != name:
            await (
                LeagueTable.update({LeagueTable.name: league.name})
                .where(LeagueTable.name == name)
                .run()
            )

        existing_league = (
            await LeagueTable.objects()
            .where(LeagueTable.name == league.name)
            .first()
            .run()
        )

        if not existing_league:
            return False

        for key, value in league.dict().items():
            setattr(existing_league, key, value)

        await existing_league.save().run()

        return True

    @staticmethod
    async def get_all() -> Iterable[League]:
        return (
            League.parse_obj(league)
            for league in await LeagueTable.select(*league_fields)
            .order_by(LeagueTable.year, ascending=False)
            .order_by(LeagueTable.name)
            .where(LeagueTable.visible == True)
            .run()
        )

    @staticmethod
    async def get_by_name(name: str) -> Optional[League]:
        return as_league(
            await LeagueTable.select(*league_fields)
            .where(LeagueTable.name == name)
            .first()
            .run()
        )

    @staticmethod
    async def delete_by_name(name: str) -> None:
        await LeagueTable.delete().where(LeagueTable.name == name).run()

    @staticmethod
    async def search(query: str) -> Iterable[League]:
        return (
            League.parse_obj(league)
            for league in await LeagueTable.select(*league_fields)
            .where(
                LeagueTable.name.ilike(f"%{query}%")
                | LeagueTable.tagline.ilike(f"%{query}%")
            )
            .order_by(LeagueTable.year, ascending=False)
            .limit(6)
            .run()
        )

    @staticmethod
    async def count() -> int:
        return int(
            (await LeagueTable.select(Count(LeagueTable.name)).first().run())["count"]
            or 0
        )


class LeagueClasses:
    @staticmethod
    async def create(league_class: LeagueClass) -> None:
        await LeagueClassTable.insert(
            LeagueClassTable(
                name=league_class.name,
                league=league_class.league,
                standard_course=league_class.standard_course,
                age_class_filter=league_class.age_class_filter,
                club_filter=league_class.club_filter,
                number_of_counting_events=league_class.number_of_counting_events,
            )
        ).run()

    @staticmethod
    async def get_by_name(league: str, cls: str) -> Optional[LeagueClass]:
        return as_league_class(
            await LeagueClassTable.select(*league_class_fields)
            .where(LeagueClassTable.league == league)
            .where(LeagueClassTable.name == cls)
            .first()
            .run()
        )

    @staticmethod
    async def get_by_league(league: str) -> Iterable[LeagueClass]:
        return (
            LeagueClass.parse_obj(league)
            for league in await LeagueClassTable.select(*league_class_fields)
            .where(LeagueClassTable.league == league)
            .order_by(LeagueClassTable.name)
            .run()
        )

    @staticmethod
    async def update(league: str, name: str, cls: LeagueClass) -> bool:
        existing_class = (
            await LeagueClassTable.objects()
            .where(LeagueClassTable.league == league)
            .where(LeagueClassTable.name == name)
            .first()
            .run()
        )

        if not existing_class:
            return False

        for key, value in cls.dict().items():
            setattr(existing_class, key, value)

        await existing_class.save().run()

        return True

    @staticmethod
    async def delete_by_name(name: str) -> None:
        await LeagueClassTable.delete().where(LeagueClassTable.name == name).run()


class LeagueGroups:
    @staticmethod
    async def get_by_league(league: str) -> dict[str, tuple[int, int]]:
        return {
            group["name"]: (group["min"], group["max"])
            for group in await LeagueGroupTable.select(*league_group_fields)
            .where(LeagueGroupTable.league == league)
            .order_by(LeagueGroupTable.name)
            .run()
        }
