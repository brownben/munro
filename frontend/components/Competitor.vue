<script setup lang="ts">
import type { PropType } from 'vue'
import {
  UserGroupIcon,
  CakeIcon,
  RectangleStackIcon,
} from '@heroicons/vue/24/outline'
import type { Competitor } from '~/api-types'

defineProps({
  competitor: { type: Object as PropType<Competitor>, required: true },
})
</script>

<template>
  <article class="flex flex-col gap-4">
    <h2
      class="text-3xl font-black text-gray-900 sm:text-4xl dark:text-gray-100"
    >
      {{ competitor.name }}
    </h2>

    <div class="flex flex-col gap-x-7 gap-y-2 sm:flex-row sm:flex-wrap">
      <ImageRow v-if="competitor.club" :icon="UserGroupIcon" darker>
        {{ competitor.club }}
      </ImageRow>
      <ImageRow v-if="competitor.age_class" :icon="CakeIcon" darker>
        {{ competitor.age_class }}
      </ImageRow>
      <ImageRow
        v-if="competitor.competitor_pool"
        :icon="RectangleStackIcon"
        darker
      >
        {{ competitor.competitor_pool }}
      </ImageRow>
    </div>

    <NuxtLink
      :to="`/competitors/${competitor.id}`"
      class="text-main-700 hover:text-main-600 dark:text-main-600 dark:hover:text-main-500 mt-3 inline-block"
    >
      View Results â†’
    </NuxtLink>
  </article>
</template>
