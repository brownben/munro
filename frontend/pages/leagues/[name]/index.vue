<script setup lang="ts">
import {
  UserIcon,
  CalculatorIcon,
  ChartBarSquareIcon,
  GlobeEuropeAfricaIcon,
} from '@heroicons/vue/24/outline/index.js'
import type { LeagueOverview } from '~/api-types'

const route = useRoute()
const loggedIn = useLoggedIn()

const { data: league, refresh } =
  await useRestrictedDataIfLoggedIn<LeagueOverview>(
    `leagues/${route.params.name}`,
    `leagues/${route.params.name}/uploadKey`
  )

const events = league.value?.events ?? []
const latestEvents = events.filter((event) => event.results_uploaded).reverse()

const todaysDate = getCurrentDate()
const nextEvent = events.filter(
  (event) => event.date >= todaysDate && !event.results_uploaded
)[0]

const scoringMethods: Record<string, [string, string]> = {
  position: ['Position Based Scoring', 'System (100 Max)'],
  position50: ['Position Based Scoring', 'System (50 Max)'],
  position99: ['Position Based Scoring', 'System (99 Max)'],
  position99average: [
    'Position Based Scoring',
    'System (99 Max, Reduced in a Draw)',
  ],
  positionDouble: ['Position Based Scoring', 'System (100 Max, Double Points)'],
  position50Double: [
    'Position Based Scoring',
    'System (50 Max, Double Points)',
  ],
  timeAverage: ['Time Relative to Average Scoring', 'System (1000 Average)'],
  timeAverage100: ['Time Relative to Average Scoring', 'System (100 Average)'],
  timeTop3: ['Time Relative to the Average Time of the Top 3', ''],
  timeTop3Adjusted: [
    'Time Relative to the Average Time of the Top 3',
    '(Adjusted by Course/Age Class)',
  ],
  timeTopAdjustedWelsh: [
    'Time Relative to Winner',
    'System (Adjusted by Course/Age Class - Welsh Multipliers)',
  ],
  positionStaggered: ['Position Based Scoring', 'System (Staggered, 60 Max)'],
  file: ['Points from File', ''],
}

if (league.value) {
  useTitle({
    title: league.value.name,
    description: `${league.value.tagline} - All the results and event information for the ${league.value.name} league on Munro`,
  })
}
</script>
<template>
  <div v-if="league">
    <Heading :title="league.name" :tagline="league.tagline">
      <ImageRow v-if="league.coordinator" :icon="UserIcon">
        Coordinated by
        <strong class="font-medium">{{ league.coordinator }}</strong>
      </ImageRow>
      <ImageRow v-if="league.scoring_method" :icon="CalculatorIcon">
        <strong class="font-medium">
          {{ scoringMethods[league.scoring_method][0] }}
        </strong>
        {{ scoringMethods[league.scoring_method][1] }}
      </ImageRow>
      <ImageRow :icon="ChartBarSquareIcon">
        Your
        <strong class="font-medium">
          best {{ league.number_of_counting_events }} scores
        </strong>
        from all
        <strong class="font-medium">
          {{ league.events.length }} events count
        </strong>
      </ImageRow>
      <ImageRow v-if="league.website" :icon="GlobeEuropeAfricaIcon" hover>
        <a :href="league.website" rel="noopener noreferrer" target="_blank">
          {{ league.website }}
        </a>
      </ImageRow>
      <template v-if="league.more_information" #extra>
        <p
          v-for="paragraph in league.more_information.split('|')"
          :key="paragraph"
          class="mt-1 text-gray-500"
        >
          {{ paragraph }}
        </p>
      </template>
    </Heading>
    <main>
      <LinksSection
        v-if="loggedIn"
        :links="[
          {
            text: 'Edit League',
            location: `/leagues/${league.name}/edit`,
          },
          {
            text: 'Create Event',
            location: `/events/create?league=${league.name}`,
          },
          {
            text: 'Manage Competitors',
            location: `/competitor-pools/${league.competitor_pool}`,
          },
        ]"
        dark
      >
        Manage your league
      </LinksSection>

      <section
        v-if="events.length >= 2 && latestEvents.length > 0"
        class="mx-auto hidden max-w-screen-lg grid-cols-1 gap-8 px-6 py-8 sm:grid sm:grid-cols-2 sm:pt-10 sm:pb-6 lg:px-8"
      >
        <div>
          <h2 class="font-bold uppercase text-gray-500 sm:pb-2">
            Latest Results
          </h2>
          <Event :event="latestEvents[0]" class="pt-4 pb-6" />
        </div>
        <div v-if="nextEvent">
          <h2 class="font-bold uppercase text-gray-500 sm:pb-2">Next Event</h2>
          <Event :event="nextEvent" class="pt-4 pb-6" />
        </div>
        <div v-else>
          <p
            class="hidden font-bold uppercase text-gray-500 sm:block sm:pb-2"
            aria-hidden="true"
          >
            &nbsp;
          </p>
          <Event
            v-if="latestEvents.length > 1"
            :event="latestEvents[1]"
            class="pt-4 pb-6"
          />
        </div>
      </section>

      <LinksSection
        v-if="league.classes.length > 0"
        :links="
          league.classes.map((course) => ({
            text: course.name,
            location: `/leagues/${league?.name}/results/${course.name}`,
          }))
        "
      >
        League Results
      </LinksSection>

      <section v-if="loggedIn">
        <div
          class="mx-auto flex max-w-screen-lg grid-cols-3 flex-col gap-8 px-6 py-6 sm:grid sm:pt-10 sm:pb-6 lg:px-8"
        >
          <div class="pb-4">
            <h2 class="col-span-1 text-2xl font-bold text-gray-500">Classes</h2>
            <NuxtLink
              :to="`/leagues/${route.params.name}/classes/create`"
              class="mt-2 inline-block rounded bg-gray-100 px-2 py-1 text-sm font-medium text-gray-700 transition hover:bg-gray-200"
            >
              + Add Class
            </NuxtLink>
          </div>

          <div class="col-span-2 -mt-6 flex flex-col divide-y">
            <LeagueClass
              v-for="cls in league.classes"
              :key="cls.name"
              :cls="cls"
              class="py-6 sm:px-4"
              @class-changed="refresh"
            />
          </div>
        </div>
      </section>

      <section
        class="mx-auto max-w-screen-lg gap-8 px-6 py-6 sm:py-12 md:py-14 lg:px-8"
      >
        <div class="flex grid-cols-3 flex-col sm:grid">
          <div class="col-span-1 pb-8 sm:pb-0">
            <h2 class="text-2xl font-bold text-gray-500">Events</h2>
            <NuxtLink
              v-if="loggedIn"
              :to="`/events/create?league=${league.name}`"
              class="mt-2 inline-block rounded bg-gray-100 px-2 py-1 text-sm font-medium text-gray-700 transition hover:bg-gray-200"
            >
              + Add Event
            </NuxtLink>
          </div>

          <div class="col-span-2 -mt-10 -mb-4 flex flex-col divide-y">
            <Event
              v-for="event in league.events"
              :key="event.id"
              :event="event"
              class="py-10 sm:px-4"
              :admin="loggedIn"
              @event-changed="refresh"
            />
          </div>
        </div>
      </section>
    </main>
  </div>
  <LazyError v-else />
</template>
