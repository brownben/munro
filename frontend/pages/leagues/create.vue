<script setup lang="ts">
import {
  IsValidURLParameter,
  IsValidURL,
  RequiredField,
} from '~/utils/validation'

const loggedIn = useLoggedIn()
if (!loggedIn.value) await redirect('/login')

const router = useRouter()

const { data: competitor_pools } = await useData<string[]>(`competitor-pools/`)

const form = reactive({
  name: '',
  tagline: '',
  year: new Date().getUTCFullYear(),
  coordinator: '',
  website: '',
  more_information: '',

  visible: true,
  scoring_method: 'position',
  number_of_counting_events: 1,
  competitor_pool: '',
})

const action = async () => {
  try {
    await usePost(`leagues/`, toRaw(form))
    await router.replace(`/leagues/${form.name}`)
  } catch (error: any) {
    if (error.data && typeof error.data?.detail === 'string')
      throw error.data.detail
    else throw 'Problem creating league, please try again.'
  }
}

useTitle({
  title: 'Create League',
  description: 'Add a new league to Munro',
})
</script>

<template>
  <div>
    <Heading title="Create League" />

    <Form button="Create" :action="action">
      <FormHeading
        title="Basic Information"
        description="The identifying data for the league."
      />

      <Input
        v-model.trim="form.name"
        label="Name:"
        class="col-span-2"
        :validator="IsValidURLParameter('a name')"
      />
      <Input
        v-model.trim="form.tagline"
        label="Description:"
        class="col-span-2"
      />
      <Input
        v-model.trim="form.coordinator"
        label="Coordinator:"
        class="col-span-2"
      />
      <InputNumber
        v-model="form.year"
        label="Year:"
        :min="1990"
        :max="2050"
        class="col-span-2"
      />
      <Input
        v-model.trim="form.website"
        label="Website: (URL)"
        type="url"
        class="col-span-2"
        :validator="IsValidURL"
      />
      <InputTextarea
        v-model.trim="form.more_information"
        label="More Information:"
        class="col-span-2"
      />

      <FormHeading
        title="Scoring"
        description="How the scores should be calculated for the league"
      />

      <InputDropdown
        v-model="form.scoring_method"
        :list="[
          { value: 'position', text: 'Position Based (100 Max)' },
          { value: 'position50', text: 'Position Based (50 Max)' },
          { value: 'position99', text: 'Position Based (99 Max)' },
          {
            value: 'position99average',
            text: 'Position Based (99 Max, Reduced in a Draw)',
          },
          {
            value: 'positionDouble',
            text: 'Position Based (100 Max, Double Points)',
          },
          {
            value: 'position50Double',
            text: 'Position Based (50 Max, Double Points)',
          },
          {
            value: 'positionStaggered',
            text: 'Position Based (Staggered, 60 Max)',
          },
          {
            value: 'timeAverage',
            text: 'Relative to Average Time (1000 Average)',
          },
          {
            value: 'timeAverage100',
            text: 'Relative to Average Time (100 Average)',
          },
          { value: 'timeTop3', text: 'Relative to Top 3 Times' },
          {
            value: 'timeTopAdjustedWelsh',
            text: 'Relative to Winner\'s Time (Welsh Multipliers)',
          },
          { value: 'file', text: 'From Upload File' },
        ]"
        label="Scoring Method:"
        :validator="RequiredField('a scoring method', true)"
        class="col-span-2"
      />
      <InputNumber
        v-model="form.number_of_counting_events"
        :validator="RequiredField('a number of counting events')"
        :min="1"
        label="Maximum Number of Counting Events:"
        class="col-span-2"
      />

      <FormHeading
        title="Additional Settings"
        description="Adjust advanced settings and enable additional features"
      />
      <InputDropdown
        v-model="form.competitor_pool"
        :list="competitor_pools?.map((name) => ({ value: name, text: name }))"
        label="Competitor Pool:"
        include-blank
        class="col-span-2"
      />
    </Form>
  </div>
</template>
