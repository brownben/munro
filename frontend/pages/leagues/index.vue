<script setup lang="ts">
import type { League } from '~/api-types'

const { data: leagues } = await useData<League[]>('leagues/')

useTitle({
  title: 'Leagues',
  description: 'All the leagues with results on Munro',
})
</script>

<template>
  <div>
    <Heading title="Leagues" />
    <main
      class="mx-auto max-w-screen-lg gap-8 px-6 py-10 sm:py-12 md:py-14 lg:px-8"
    >
      <div
        class="col-span-2 -mx-4 -mb-4 -mt-8 flex flex-col divide-y dark:divide-gray-600"
      >
        <League
          v-for="league in leagues"
          :key="league.name"
          :league="league"
          class="px-4 py-10"
        />
      </div>
    </main>
  </div>
</template>
