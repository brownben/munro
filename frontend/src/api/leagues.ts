import { getData, postData, putData, deleteData } from './requests'

export const getLeague = (name: string): Promise<League | null> =>
  getData<League | null>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Fetching League Details',
  })

export const getLeagues = (): Promise<League[] | null> =>
  getData<League[]>({
    apiLocation: '/api/leagues',
    customErrorMessage: 'Problem Fetching Leagues',
  })

export const createLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/leagues`,
    data,
    customErrorMessage: 'Problem Creating League',
    customSuccessMessage: `League \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/leagues/${data.name}`,
    data,
    customErrorMessage: 'Problem Updating League',
    customSuccessMessage: `League \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteLeague = (name: string): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Deleting League',
    customErrorHandler: true,
  })
