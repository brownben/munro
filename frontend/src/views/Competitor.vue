<template>
  <div>
    <vue-headful
      :title="`Munro - ${competitor.name || ''} - Competitor`"
      :description="`Results for ${competitor.name || ''} in the ${
        competitor.league || ''
      } league on Munro - League Results. Sorted. Sports League Results Calculated Quick and Easily, with Results Sorting and Filtering Options`"
      :url="`https://munro-leagues.herokuapp.com/competitors/${$route.params.id}`"
      :head="{
        meta: { name: 'robots', content: 'all' },
      }"
    />

    <Layout v-if="competitor && competitor.name" hasMobileSubTitle>
      <template #title>
        <router-link
          :to="`/leagues/${competitor.league}`"
          class="mb-1 text-xl font-bold text-main-700 font-heading"
        >
          {{ competitor.league }}
        </router-link>
        <h1 class="text-4xl font-bold font-heading leading-12">
          {{ competitor.name }}
        </h1>
        <h2 class="text-xl text-gray-600 font-heading">
          <span class="mr-2 md:mr-4">{{ competitor.course }}</span>
          <span v-if="competitor.club" class="mr-2 md:mr-4">
            {{ competitor.club }}</span
          >
          <span v-if="competitor.ageClass"> {{ competitor.ageClass }}</span>
        </h2>
      </template>

      <div
        v-if="competitor && $auth.user"
        class="col-span-2 card card-color-dark"
      >
        <h2 class="text-3xl font-bold text-white font-heading">
          Admin Actions
        </h2>

        <div>
          <router-link
            :to="`/competitors/${$route.params.id}/edit`"
            class="button button-white"
            >Edit Competitor</router-link
          >
          <router-link to="/competitors/merge" class="button button-white"
            >Merge Competitors</router-link
          >
          <router-link to="/results/transfer" class="button button-white"
            >Transfer Result</router-link
          >
          <router-link to="/results/manual" class="button button-white"
            >Manual Points</router-link
          >
        </div>
      </div>

      <ResultOverviewCard
        v-for="result of results"
        :key="result.id"
        :result="result"
        @resultChanged="getCompetitorResults"
      />
    </Layout>
    <not-found v-if="!competitor" />
  </div>
</template>

<script>
import axios from 'axios'

import Layout from '@/components/Layout.vue'
import ResultOverviewCard from '@/components/cards/ResultOverviewCard.vue'

const NotFound = () => import('@/views/NotFound.vue')

export default {
  components: {
    Layout,
    ResultOverviewCard,
    NotFound,
  },

  data: function () {
    return {
      competitor: {},
      results: [],
      auth: this.$auth,
    }
  },

  watch: {
    $route: {
      immediate: true,
      handler: function () {
        this.getCompetitor()
        this.getCompetitorResults()
      },
    },
  },

  methods: {
    getCompetitor: function () {
      return axios
        .get(`/api/competitors/${this.$route.params.id}`)
        .then((response) => {
          this.competitor = response.data
        })
        .catch(() =>
          this.$messages.addMessage('Problem Getting Competitor Details')
        )
    },

    getCompetitorResults: function () {
      return axios
        .get(`/api/competitors/${this.$route.params.id}/results`)
        .then((response) => {
          this.results = response.data
        })
        .catch(() =>
          this.$messages.addMessage('Problem Getting Competitor Results')
        )
    },
  },
}
</script>
