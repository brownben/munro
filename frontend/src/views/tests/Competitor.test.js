import axios from 'axios'
import { mount, shallowMount } from '@vue/test-utils'
import Competitor from '@/views/Competitor'
import { sampleSingleLeague } from '@/tests/test data/leagues'

jest.mock('axios')
const flushPromises = () =>
  new Promise((resolve) => global.setImmediate(resolve))

beforeEach(() => {
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: [] })
})

test('Is a Vue Instance', () => {
  const wrapper = mount(Competitor, {
    mocks: {
      $route: { path: '', params: {} },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = mount(Competitor, {
    mocks: {
      $route: { path: '', params: { id: '' } },
      $auth: { user: true },
    },
    propsData: {
      competitor: {
        name: 'Bob Jones',
        club: 'FVO',
        league: 'Sprintelope 2019',
        ageClass: 'M50',
      },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Get Competitor - Correct API Call', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(Competitor, {
    mocks: {
      $auth: { user: false },
      $route: { params: { id: 'Test' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue()
  await wrapper.vm.getCompetitor()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/competitors/Test')
})

test('Get Competitor - Success', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(Competitor, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: sampleSingleLeague[0] })
  await wrapper.vm.getCompetitor()
  expect(mockAddMessage).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.competitor).toEqual(sampleSingleLeague[0])
})

test('Get Competitor - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(Competitor, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getCompetitor()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith(
    'Problem Getting Competitor Details'
  )
  wrapper.setData({ competitor: false })
  expect(wrapper.element).toMatchSnapshot()
})

test('Get Competitor Results - Correct API Call', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(Competitor, {
    mocks: {
      $auth: { user: false },
      $route: { params: { id: 'Test' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue()
  await wrapper.vm.getCompetitorResults()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/competitors/Test/results')
})

test('Get Competitor Results - Success', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(Competitor, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: sampleSingleLeague[0] })
  await wrapper.vm.getCompetitorResults()
  expect(mockAddMessage).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.results).toEqual(sampleSingleLeague[0])
})

test('Get Competitor Results - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(Competitor, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getCompetitorResults()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith(
    'Problem Getting Competitor Results'
  )
})
