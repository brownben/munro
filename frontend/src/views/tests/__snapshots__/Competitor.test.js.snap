// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get Competitor - Error 1`] = `<!---->`;

exports[`Renders Correctly 1`] = `<!---->`;
